aten/src/ATen/Parallel.h

```diff
@@ -32,6 +33,19 @@ CAFFE2_API int get_thread_num();
 // Checks whether the code runs in parallel region
 CAFFE2_API bool in_parallel_region();
 
+namespace internal {
+
+// Initialise num_threads lazily at first parallel call
+inline CAFFE2_API void lazy_init_num_threads() {
+  thread_local bool init = false;
+  if (C10_UNLIKELY(!init)) {
+    at::init_num_threads();
+    init = true;
+  }
+}
+
```

Comment:
I'm a bit concerned about multiple compilation units including this function in the header, could we move the implementation into the .cpp ? e.g. ATen/ParallelCommon.cpp

Reply:
My intention is that the cheap boolean check will be inlined into the caller and in the overwhelmingly likely case, there is no function call overhead.

Reply:
Also, it's marked `inline` so duplicates in multiple compilation units will be deduplicated at link time.